kvs/client/python/anna/ipc_client.py

```diff
@@ -91,9 +199,57 @@ def put(self, key, value):
         else:
             raise ValueError('Invalid PUT type: ' + str(type(value)))
 
-        self.put_socket.send(request.SerializeToString())
+        request.response_address = PUT_RESPONSE_ADDR_TEMPLATE % thread_id
 
-        resp = KeyResponse()
-        resp.ParseFromString(self.put_socket.recv())
+        self.put_push_socket.send(request.SerializeToString())
 
-        return resp.tuples[0].error == 0
+        try:
+            msg = self.put_pull_socket.recv()
+        except zmq.ZMQError as e:
+            if e.errno == zmq.EAGAIN:
+                print("request timed out")
+                return False
+            else:
+                print("unknown zmq error")
+                return False
+        else:
+            resp = KeyResponse()
+            resp.ParseFromString(msg)
+
+            return resp.tuples[0].error == 0
+
+    def causal_put(self, key, vector_clock, dependency, value, client_id):
+        request = CausalRequest()
+        request.consistency = CROSS
+        request.id = client_id
+
+        tp = request.tuples.add()
+        tp.key = key
+
+        cross_causal_value = CrossCausalValue()
+        cross_causal_value.vector_clock = vector_clock
+
+        for key in dependency:
+            dep = cross_causal_value.deps.add()
+            dep.key = key
+            dep.vector_clock = dependency[key]
+
+        cross_causal_value.values.add(value)
+
+        tp.payload = SerializeToString(cross_causal_value)
+
+        request.response_address = PUT_RESPONSE_ADDR_TEMPLATE % thread_id
```

Comment:
This is the same on every request, right? We should just create it once and reuse it.

Reply:
fixed